src/Common/src/Interop/Unix/System.Native/Interop.readdir.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        internal enum NodeType : short
+        {
+            Unknown         = 0,
+            NamedPipe       = 1,
+            CharacterDevice = 2,
+            Directory       = 4,
+            BlockDevice     = 6,
+            File            = 8,
+            SymLink         = 10,
+            Socket          = 12,
+            BsdWhiteout     = 14,
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal unsafe struct DirectoryEntry
+        {
+            internal ushort      reclen;
+            internal ulong       inode_number;
+            internal ulong       offset;
+            internal NodeType    inode_type;
+            internal fixed byte  name[1024];
```

Comment:
In general, I think it's fine to limit things to what we need when there's an obvious perf win. That would also give us tighter code with fewer things to configure/ifdef, which is another fine reason to leave parts out IMHO. Also, someof the fields we're not using are not mentioned in POSIX, which is another reason to exclude.

However, as far as returning pointers without copying, I think we're currently on shaky ground. POSIX says readdir is not thread-safe and we should probably be using readdir_r. Now might be a good time to build something more reuasable for the various _r buffers. Possibly with pooling and less chattiness on resize....